File: Graph/ShortestPath/dijkstra.cpp
#include <bits/stdc++.h>
using namespace std;
template <typename T>
class Graph {
    unordered_map<T, list<pair<T, int>>> g;

public:
    void addEdge(T u, T v, int dist, bool isBir = true)
    {
        g[u].push_back({ v, dist });
        if (isBir)
            g[v].push_back({ u, dist });
    }
    void printAdj()
    {
        //iterate over all the keys in map
        for (auto j : g) {
            cout << j.first << " ->";
            for (auto l : j.second) {
                cout << " (" << l.first << ", " << l.second << ")";
            }
            cout << endl;
        }
    }
    void disjkstraSSSP(T src)
    {
        unordered_map<T, int> dist;
        for (auto j : g) {
            dist[j.first] = INT_MAX;
        }
        //make a set to find a node with minimum distance
        set<pair<int, T>> s;
        dist[src] = 0;
        s.insert({ 0, src });
        while (!s.empty()) {
            //find the pair at the front
            auto p = *(s.begin());
            T node = p.second;
            int nodeDist = p.first;
            s.erase(s.begin());
            //iterate over neighbours of current node
            for (auto childPair : g[node]) {
                if (nodeDist + childPair.second < dist[childPair.first]) {
                    //in the set update is not possible , so
                    //we have to remove the pair and then insert again
                    T dest = childPair.first;
                    auto f = s.find({ dist[dest], dest });
                    if (f != s.end()) {
                        s.erase(f);
                    }
                    //insert the new pair
                    dist[dest] = nodeDist + childPair.second;
                    s.insert({ dist[dest], dest });
                }
            }
        }
        //print dist of all other node from src
        for (auto d : dist) {
            cout << d.first << " is located at a distance of " << d.second << " from " << src << endl;
        }
    }
};
int main()
{
    Graph<string> india;
    india.addEdge("Amritsar", "Delhi", 1);
    india.addEdge("Amritsar", "Jaipur", 4);
    india.addEdge("Jaipur", "Delhi", 2);
    india.addEdge("Jaipur", "Mumbai", 8);
    india.addEdge("Bhopal", "Agra", 2);
    india.addEdge("Mumbai", "Bhopal", 3);
    india.addEdge("Agra", "Delhi", 1);
    // india.printAdj();
    india.disjkstraSSSP("Amritsar");
}

File: Graph/ShortestPath/Dijkstra.java
package Graph.ShortestPath;

import java.util.*;

class Pair<T> {
    T first;
    int second;

    Pair() {
    }

    Pair(T first, int second) {
        this.first = first;
        this.second = second;
    }
}

public class Dijkstra {
    public static void main(String[] args) {
        Graph<String> india = new Graph<>();
        india.addEdge("Amritsar", "Delhi", 1);
        india.addEdge("Amritsar", "Jaipur", 4);
        india.addEdge("Jaipur", "Delhi", 2);
        india.addEdge("Jaipur", "Mumbai", 8);
        india.addEdge("Bhopal", "Agra", 2);
        india.addEdge("Mumbai", "Bhopal", 3);
        india.addEdge("Agra", "Delhi", 1);
        india.disjkstraSSSP("Amritsar");
    }

    static class Graph<T> {
        Map<T, List<Pair<T>>> g = new HashMap<>();

        void addEdge(T src, T dest, int dist) {
            if (!g.containsKey(src)) {
                g.put(src, new ArrayList<>());
            }
            List<Pair<T>> ll = g.get(src);
            ll.add(new Pair<T>(dest, dist));
            g.put(src, ll);
            if (!g.containsKey(dest)) {
                g.put(dest, new ArrayList<>());
            }
            List<Pair<T>> lld = g.get(dest);
            lld.add(new Pair<T>(src, dist));
            g.put(dest, lld);
        }

        void disjkstraSSSP(T src) {
            Map<T, Integer> dist = new HashMap<>();
            g.keySet().forEach(key -> dist.put(key, Integer.MAX_VALUE));
            TreeSet<Pair<T>> s = new TreeSet<>((a, b) -> a.second - b.second);
            s.add(new Pair<>(src, 0));
            dist.put(src, 0);
            while (!s.isEmpty()) {
                var p = s.first();
                var node = p.first;
                int nodeDist = p.second;
                s.remove(p);
                for (var childPair : g.get(node)) {
                    var dest = childPair.first;
                    if (nodeDist + childPair.second < dist.get(dest)) {
                        s.remove(new Pair<>(dest, dist.get(dest)));
                        dist.put(dest, nodeDist + childPair.second);
                        s.add(new Pair<>(dest, dist.get(dest)));
                    }
                }
            }
            dist.entrySet().forEach(entry -> System.out
                    .println(entry.getKey() + " is located at a distance of " + entry.getValue() + " from " + src));
        }
    }
}